### PR TITLE
315 batch delivery

### PR DESCRIPTION
**Pull request by wojtkiewicz**
Batch delivery introduced. 

**Feature switch:**
Enabled by creating subscription with deliveryType == BATCH.

**Bounded by 3 main factors:**
- batchSize (number of messages in a batch)
- batchTime (max period of time between first message in a batch and delivery)
- batchVolume (max number of bytes in a batch)

**Supported content types:**
Right now only *subscription.contentType* == JSON is supported but for *topic.contentType* we support both AVRO and JSON.

**Message metadata:**
We send *Hermes-Batch-Id* in a header to subscriber. Each message in a batch is wrapped with its own metadata (for instance containing *Hermes-Message-Id* or *Trace-Id*). Details are very content type specific but for a JSON delivery we send an array of wrapped original messages.

**Timeouts:**
Can be configured for each subscription individually.

**Resources:**
Entire batch processing is single-threaded. We reuse the same thread (one per subscription) that was previously responsible for message consumption from broker. Batches are aggregated off-heap with a memory bounded DirectBufferPool.

**Performance impact:**
It was not measured yet but because we store batches off-heap GC should not suffer too much. Memory is bounded and maintained by DirectBufferPool so in case of memory exhaustion we should block on access and not blow up with OutOfMemoryException. Messages that are too large for a configured batchVolume are discarded. For performance reasons conversion between AVRO and JSON should be avoided. Right now this option is designed as a migration utility.

**Retry policy:**
We use fixed backoff policy with a maximum TTL. Both values can be configured individually for every subscription. Batch is considered discarded if retry policy was exhausted.

**Metrics:**
Metrics are translated for existing counters/meters so for instance inflight meter will show how many messages are currently in a batch and delivered/discarded meters will increase by the size of entire batch.

**Tracking:**
Added new field with batch id. API didn't change from the client perspective. Timestamps and statuses are the same for all messages in a given batch.

**Subscription model compatibility:**
Changes to subscription model are backwards compatible both from API and from ZK perspective. Worth mentioning is that for ZK there is a thin migration layer that converts old subscription policy to SERIAL. This migration layer does not modify data in ZK when encounters old subscription policy but does it on the fly. But beware that when you do any update operation on Subscription then new subscription policy model will be used.

**Docs:**
In progress.

### CODE REVIEW

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                    offsets.putAllDelivered(batch.getPartitionOffsets());
+                });
+                result.getDiscarded().forEach(m -> monitoring.markDiscarded(m, subscription, "too large"));
+            } finally {
+                logger.debug("Cleaning batch for subscription {}", subscription.getId());
+                inflight.ifPresent(this::clean);
+            }
+        }
+        logger.info("Stopped consumer for subscription {}", subscription.getId());
+        unsetThreadName();
```
Comment (adamdubiel) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: shouldnt these go into finally?

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    private Retryer<MessageSendingResult> createRetryer(MessageBatch batch, BatchSubscriptionPolicy policy) {
+        return createRetryer(batch, policy.getMessageBackoff(), policy.getMessageTtl(), policy.isRetryClientErrors());
+    }
+
+    private Retryer<MessageSendingResult> createRetryer(final MessageBatch batch, int messageBackoff, int messageTtl, boolean retryClientErrors) {
+        return RetryerBuilder.<MessageSendingResult>newBuilder()
+                .retryIfExceptionOfType(IOException.class)
+                .retryIfRuntimeException()
+                .retryIfResult(result -> isConsuming() && !result.succeeded() && (!result.isClientError() || retryClientErrors))
```
Comment (adamdubiel) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: maybe add some labels to this? its hard to parse the exact conditions

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+                .withRetryListener(getRetryListener(result -> monitoring.markFailed(batch, subscription, result)))
+                .build();
+    }
+
+    private void deliver(MessageBatch batch, Retryer<MessageSendingResult> retryer) {
+        try (Timer.Context timer = hermesMetrics.subscriptionLatencyTimer(subscription).time()) {
+            MessageSendingResult result = retryer.call(() -> sender.send(batch, subscription.getEndpoint(), subscription.getBatchSubscriptionPolicy().getRequestTimeout()));
+            monitoring.markSendingResult(batch, subscription, result);
+        } catch (Exception e) {
+            logger.error(format("[batch_id=%s, subscription=%s] Batch was rejected.", batch.getId(), subscription.toSubscriptionName()), e);
```
Comment (adamdubiel) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: another log format? maybe make it compatible with existing? (or change existing)

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/batch/ByteBufferMessageBatchFactory.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Override
+    public MessageBatch createBatch(Subscription subscription) {
+        try {
+            ByteBuffer buffer = bufferPool.allocate(subscription.getBatchSubscriptionPolicy().getBatchVolume());
+            switch (subscription.getContentType()) {
+                case JSON:
+                    return new JsonMessageBatch(randomUUID().toString(), buffer, subscription, clock);
+                case AVRO:
```
Comment (adamdubiel) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/batch/ByteBufferMessageBatchFactory.java]: do we have any idea how to support AVRO in the future?

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/JettyMessageSender.java [hunk 5]
```diff
@@ -50,12 +43,11 @@ protected void sendMessage(Message message, final CompletableFuture<MessageSendi
     }
 
     private Request buildRequest(Message message) throws EndpointAddressResolutionException {
-
         Request request = client.newRequest(endpoint.resolveFor(message))
                 .method(HttpMethod.POST)
                 .header(HttpHeader.KEEP_ALIVE.toString(), "true")
                 .header(MESSAGE_ID.getName(), message.getId())
-                .header(HttpHeader.CONTENT_TYPE.toString(), contentTypeToMediaType.apply(message.getContentType()))
+                .header(HttpHeader.CONTENT_TYPE.toString(), MediaType.APPLICATION_JSON)
```
Comment (adamdubiel) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/JettyMessageSender.java]: why JSON is hardcoded here?

File: hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/ConsumerTest.java [hunk 6]
```diff
@@ -147,6 +151,7 @@ public void shouldKeepReadingMessagesAfterTimeout() {
     }
 
     @Test
+    @Ignore
```
Comment (adamdubiel) [inline hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/ConsumerTest.java]: hmm... lots of ignores?

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/batch/ByteBufferMessageBatchFactory.java (hunk 4 above)
Comment (wojtkiewicz) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/batch/ByteBufferMessageBatchFactory.java]: We need to do some research to check whether there is a better solution other than sending message with a batch schema that contains list of events with message schema

File: hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/batch/DirectBufferPoolTest.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.junit.Assert.*;
+
+public class DirectBufferPoolTest {
```
Comment (adamdubiel) [inline hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/batch/DirectBufferPoolTest.java]: somne test written in spock, some in junit in same PR - why?

File: hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/ByteBufferEntityTest.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+import org.junit.Test;
+
+import java.io.ByteArrayOutputStream;
+import java.nio.ByteBuffer;
+
+/**
+ * Unit tests for {@link ByteBufferEntity}.
+ *
+ */
+public class ByteBufferEntityTest {
```
Comment (adamdubiel) [inline hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/ByteBufferEntityTest.java]: all "copied" classes should have this "i was copied" remark

File: hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/supervisor/workload/SelectiveSupervisorControllersIntegrationTest.java [hunk 9]
```diff
@@ -14,6 +15,7 @@
 import static org.assertj.core.api.Assertions.assertThat;
 import static pl.allegro.tech.hermes.test.helper.endpoint.TimeoutAdjuster.adjust;
 
+@Ignore
```
Comment (adamdubiel) [inline hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/supervisor/workload/SelectiveSupervisorControllersIntegrationTest.java]: ?

File: integration/src/integration/java/pl/allegro/tech/hermes/integration/BatchDeliveryTest.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+        // then
+        expectSingleBatch(SMALL_BATCH);
+    }
+
+    @Test
+    public void shouldDeliverBatchInGivenTimePeriod() throws IOException {
+        // given
+        Topic topic = operations.buildTopic("deliverBatchInGivenTimePeriod", "topic");
+        operations.createBatchSubscription(topic, HTTP_ENDPOINT_URL, LARGE_MESSAGE_TTL, LARGE_BATCH_SIZE, SMALL_BATCH_TIME, LARGE_BATCH_VOLUME);
```
Comment (adamdubiel) [inline integration/src/integration/java/pl/allegro/tech/hermes/integration/BatchDeliveryTest.java]: repeating the same BIG_VALUES doesnt make it more readable.. either make a builder for batch subscription or aply some defaults, or just use values instead of consts

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java (hunk 1 above)
Comment (wojtkiewicz) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: good idea, done

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java (hunk 2 above)
Comment (wojtkiewicz) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: I added label for parsing client error

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/JettyMessageSender.java (hunk 5 above)
Comment (wojtkiewicz) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/JettyMessageSender.java]: no idea, reverted to original

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java (hunk 3 above)
Comment (wojtkiewicz) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: introduced format "log message [application state]"

File: hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/ConsumerTest.java (hunk 6 above)
Comment (wojtkiewicz) [inline hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/ConsumerTest.java]: There were problems with this test after I introduced interfaces for Consumer and I didn't have patience to trace all mocking issues because of their excessive usage in this test. Anyway, it is working now.

File: hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/batch/DirectBufferPoolTest.java (hunk 7 above)
Comment (wojtkiewicz) [inline hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/batch/DirectBufferPoolTest.java]: I don't have good answer for this. I can rewrite that spock test in junit for the sake of consistency if you want...

File: hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/ByteBufferEntityTest.java (hunk 8 above)
Comment (wojtkiewicz) [inline hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/ByteBufferEntityTest.java]: Added explanation in a comment why and from where this class comes from.

File: hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/supervisor/workload/SelectiveSupervisorControllersIntegrationTest.java (hunk 9 above)
Comment (wojtkiewicz) [inline hermes-consumers/src/test/java/pl/allegro/tech/hermes/consumers/supervisor/workload/SelectiveSupervisorControllersIntegrationTest.java]: enabled

File: integration/src/integration/java/pl/allegro/tech/hermes/integration/BatchDeliveryTest.java (hunk 10 above)
Comment (wojtkiewicz) [inline integration/src/integration/java/pl/allegro/tech/hermes/integration/BatchDeliveryTest.java]: At the time of writing this test we didn't have any builders ready for creating batch subscriptions. I think using BatchSubscriptionPolicy.Builder here should be sufficient to address readability.

Comment (wojtkiewicz) [inline integration/src/integration/java/pl/allegro/tech/hermes/integration/BatchDeliveryTest.java]: I refactored this test to be more readable (I hope).

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/BatchSubscriptionPolicy.java [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+                    subscriptionPolicy.batchVolume);
+        }
+
+        public Builder applyDefaults() {
+            subscriptionPolicy.messageTtl = 60 * 1000;
+            subscriptionPolicy.messageBackoff = 500;
+            subscriptionPolicy.requestTimeout = 30 * 1000;
+            subscriptionPolicy.batchSize = 100;
+            subscriptionPolicy.batchTime = 30 * 1000;
+            subscriptionPolicy.batchVolume = 64 * 1000;
```
Comment (dankraw) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/BatchSubscriptionPolicy.java]: would be nice if these values were configurable.

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+    private boolean shouldRetryOnClientError(boolean retryClientErrors, MessageSendingResult result) {
+        return !result.isClientError() || retryClientErrors;
+    }
+
+    private void deliver(MessageBatch batch, Retryer<MessageSendingResult> retryer) {
+        try (Timer.Context timer = hermesMetrics.subscriptionLatencyTimer(subscription).time()) {
+            MessageSendingResult result = retryer.call(() -> sender.send(batch, subscription.getEndpoint(), subscription.getBatchSubscriptionPolicy().getRequestTimeout()));
+            monitoring.markSendingResult(batch, subscription, result);
+        } catch (Exception e) {
+            logger.error(format("Batch was rejected [batch_id=%s, subscription=%s].", batch.getId(), subscription.toSubscriptionName()), e);
```
Comment (dankraw) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: string.format could be omitted, and %s -> {}

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/ApacheMessageBatchSender.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+import pl.allegro.tech.hermes.consumers.consumer.sender.MessageBatchSender;
+import pl.allegro.tech.hermes.consumers.consumer.sender.MessageSendingResult;
+
+import java.io.IOException;
+
+import static pl.allegro.tech.hermes.api.ContentType.AVRO;
+import static pl.allegro.tech.hermes.common.http.MessageMetadataHeaders.BATCH_ID;
+import static pl.allegro.tech.hermes.consumers.consumer.sender.http.AvroMediaType.AVRO_BINARY;
+
+public class ApacheMessageBatchSender implements MessageBatchSender {
```
Comment (dankraw) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/ApacheMessageBatchSender.java]: ApacheHttpClientMessageBatchSender would be more appropriate :P

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/SubscriptionPolicy.java [hunk 14]
```diff
@@ -25,8 +25,8 @@
     private SubscriptionPolicy() { }
 
     @JsonCreator
-    public SubscriptionPolicy(@JsonProperty("rate") int rate, @JsonProperty("messageTtl") int messageTtl,
-                              @JsonProperty("retryClientErrors") boolean retryClientErrors,
+    public SubscriptionPolicy(@JsonProperty("rate") Integer rate, @JsonProperty("messageTtl") Integer messageTtl,
```
Comment (dankraw) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/SubscriptionPolicy.java]: can we rename this class to SerialSubscriptionPolicy for the sake of clarity?

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java (hunk 12 above)
Comment (wojtkiewicz) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: not really if you pass an exception

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/BatchSubscriptionPolicy.java (hunk 11 above)
Comment (wojtkiewicz) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/BatchSubscriptionPolicy.java]: server side?

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java (hunk 12 above)
Comment (dankraw) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: This is actually supported by slf4j:
```java
logger.error("Batch was rejected [batch_id={}, subscription={}].", batch.getId(), subscription.toSubscriptionName(), e);
```

Comment (wojtkiewicz) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: are you sure? you are calling *public void error(String format, Object... arguments)* now

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/BatchSubscriptionPolicy.java (hunk 11 above)
Comment (dankraw) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/BatchSubscriptionPolicy.java]: ok, it's an api object. nevermind :>

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java (hunk 12 above)
Comment (dankraw) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: yep. try it.

Comment (dankraw) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: http://stackoverflow.com/questions/6371638/slf4j-how-to-log-formatted-message-object-array-exception

Comment (wojtkiewicz) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: wtf... I will try this but our logback backend was putting null in place of throwable when I looked up sources...

Comment (adamdubiel) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: http://allegro.tech/2015/10/there-is-more-to-logging-than-meets-the-eye.html :P

Comment (wojtkiewicz) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: I checked and it works but to me it really shouldn't... Don't understand why can't pass an exception before varargs so the whole thing isn't so obscure...

Comment (wojtkiewicz) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/BatchConsumer.java]: Done

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/ApacheMessageBatchSender.java (hunk 13 above)
Comment (wojtkiewicz) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/http/ApacheMessageBatchSender.java]: Done

**Comment by dankraw** [comment]
:+1:

File: hermes-api/src/main/java/pl/allegro/tech/hermes/api/SubscriptionPolicy.java (hunk 14 above)
Comment (wojtkiewicz) [inline hermes-api/src/main/java/pl/allegro/tech/hermes/api/SubscriptionPolicy.java]: discussed offline

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/batch/ByteBufferMessageBatchFactory.java (hunk 4 above)
Comment (adamdubiel) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/batch/ByteBufferMessageBatchFactory.java]: i think its a decent solution actually - send the schema

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/resolver/SimpleEndpointAddressResolver.java [hunk 15]
```diff
@@ -0,0 +1,6 @@
+package pl.allegro.tech.hermes.consumers.consumer.sender.resolver;
+
+
+public class SimpleEndpointAddressResolver implements EndpointAddressResolver {
```
Comment (adamdubiel) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/resolver/SimpleEndpointAddressResolver.java]: why do i need this "Simple" class?

**Comment by adamdubiel** [comment]
this pr is too big, what happened to "small prs to mother branch"? you added discovery support as 3 commits, i cant even review that code without going through 84 file changes

**Comment by adamdubiel** [comment]
:+1: as 'incubating'

File: hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/resolver/SimpleEndpointAddressResolver.java (hunk 15 above)
Comment (wojtkiewicz) [inline hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/sender/resolver/SimpleEndpointAddressResolver.java]: This is the most basic form of address resolver which takes as an input only actual address. This is especially useful for test scenarios and batch delivery.

**Comment by wendigo** [comment]
Does it support also avro?

**Comment by wojtkiewicz** [comment]
**Supported content types:**
Right now only subscription.contentType == JSON is supported but for topic.contentType we support both AVRO and JSON.

**Comment by wendigo** [comment]
Is it possible to support also avro subscription type?

**Comment by wojtkiewicz** [comment]
Next iteration :-)
